src/solvers/refinement/string_constraint.h

```diff
@@ -0,0 +1,115 @@
+
+#ifndef CPROVER_SOLVER_STRING_CONSTRAINT_H
+#define CPROVER_SOLVER_STRING_CONSTRAINT_H
+
+#include <langapi/language_ui.h>
+#include <solvers/refinement/bv_refinement.h>
+#include <solvers/refinement/refined_string_type.h>
+
+class string_constraintt : public exprt
+{
+private:
+  // String axioms can have 4 different forms:
+  // either a simple expression p,
+  // or a string constant: forall x in [0,|s|[. s(x) = c(x)
+  // or universally quantified expression: forall x in [lb,ub[. p(x)
+  // or a expression for non containment:
+  // forall x in [lb,ub[. p(x) => exists y in [lb,ub[. s1[x+y] != s2[y]
+  enum {SIMPLE, STRING_CONSTANT, UNIV_QUANT, NOT_CONTAINS} form;
+
+  // Universally quantified symbol
+  symbol_exprt quantified_variable;
+
+  // Bounds on the quantified variables (alternate between inf and sup)
+  std::vector<exprt> bounds;
+
+  // Only for NOT_CONTAINS constraints (represent s1 and s2)
+  std::vector<exprt> compared_strings;
+
+  // we should change the structure of the class to adopt the notations of chapter 7
+  //  exprt index_guard;
+
+public:
+
+  // used to store information about witnesses for not_contains constraints
+  symbol_exprt witness;
+
+
+  // True axiom
+  string_constraintt() : exprt(true_exprt()) { form = SIMPLE; }
+
+  // Axiom with no quantification, and no premise
+  string_constraintt(exprt bod, bool is_string_constant=false) : exprt(bod) { form = is_string_constant?SIMPLE:STRING_CONSTANT; }
+
+  // Axiom with no quantification: prem => bod
+  string_constraintt(exprt prem, exprt bod)  : exprt(implies_exprt(prem,bod))
+  { form = SIMPLE; }
+
+  // Add an universal quantifier
+  string_constraintt forall(const symbol_exprt & univ, const exprt & bound_inf, const exprt & bound_sup);
+  // Default bound inferior is 0
+  string_constraintt forall(const symbol_exprt & univ, const exprt & bound_sup);
+
+  // Bound a variable that is existentially quantified
+  string_constraintt exists(const symbol_exprt & exist, const exprt & bound_inf, const exprt & bound_sup);
+  // Default bound inferior is 0
+  string_constraintt exists(const symbol_exprt & exist, const exprt & bound_sup);
+
+  static string_constraintt not_contains
+  (exprt univ_lower_bound, exprt univ_bound_sup, exprt premise,
+   exprt exists_bound_inf, exprt exists_bound_sup, exprt s0, exprt s1);
+
+  bool is_simple() const { return (form == SIMPLE); };
+  bool is_string_constant() const { return (form == STRING_CONSTANT); };
+  bool is_univ_quant() const { return (form == UNIV_QUANT); };
+  bool is_not_contains() const { return (form == NOT_CONTAINS); };
+
+  exprt premise() const;
+
+  exprt body() const;
+
+  inline exprt s0() const { assert(is_not_contains()); return compared_strings[0];}
+  inline exprt s1() const { assert(is_not_contains()); return compared_strings[1];}
+
+
+  inline symbol_exprt get_univ_var() const { assert(form==UNIV_QUANT); return quantified_variable;}
```

Comment:
Ids introduced in https://github.com/diffblue/cbmc/pull/341/commits/0f6409cb740d5d2f9cab03cbd959e1b52a875a60